Source/ARTLocalDevice.m

```diff
@@ -72,24 +71,18 @@ + (ARTLocalDevice *)load:(NSString *)clientId storage:(id<ARTDeviceStorage>)stor
     #endif
     device.push.recipient[@"transportType"] = ARTDevicePushTransportType;
 
-    NSString *deviceId = [storage objectForKey:ARTDeviceIdKey];
+    NSString *deviceId = [storage objectForKey:ARTDeviceIdKey] ?: @""; // PCD2, not nullable
     NSString *deviceSecret = deviceId == nil ? nil : [storage secretForDevice:deviceId];
     
-    if (deviceId == nil || deviceSecret == nil) { // generate both at the same time
-        deviceId = [self generateId];
-        deviceSecret = [self generateSecret];
-        
-        [storage setObject:deviceId forKey:ARTDeviceIdKey];
-        [storage setSecret:deviceSecret forDevice:deviceId];
-    }
-    
     device.id = deviceId;
     device.secret = deviceSecret;
 
     id identityTokenDetailsInfo = [storage objectForKey:ARTDeviceIdentityTokenKey];
     ARTDeviceIdentityTokenDetails *identityTokenDetails = [ARTDeviceIdentityTokenDetails unarchive:identityTokenDetailsInfo withLogger:logger];
     device->_identityTokenDetails = identityTokenDetails;
 
+    device.clientId = identityTokenDetails.clientId;
```

Comment:
I don't think this matches the behaviour of the spec — my understanding of [RSH8a](https://sdk.ably.com/builds/ably/specification/main/features/#RSH8a), [RSH8b](https://sdk.ably.com/builds/ably/specification/main/features/#RSH8b), and [RSH8d](https://sdk.ably.com/builds/ably/specification/main/features/#RSH8d) is that `clientId` is meant to be persisted as a standalone attribute. (Again, I'd suggest checking how Android does it, too.) We also need to update the tests for RSH8d to make sure that the updated value gets persisted.

(For existing installations of the SDK we will however need to also populate this persisted value from the existing data, and I guess that we can use `identityTokenDetails.clientId` for that.)

Reply:
Android [uses](https://github.com/ably/ably-java/blob/72992c943498ee2ebefb8acec303718ed7f35ae4/android/src/main/java/io/ably/lib/push/LocalDevice.java#L64) a separate attribute for `clientId`, but I would vote against it, how does it make any sense?

Reply:
What do you mean by "it doesn't make any sense"? Do you not agree that it's the specified behaviour? Bear in mind that the `LocalDevice`’s `clientId` doesn't necessarily match that of the currently saved `identityTokenDetails` (see e.g. [RSH8e](https://sdk.ably.com/builds/ably/specification/main/features/#RSH8e), where `LocalDevice.clientId` has been updated but the registration has not).

Reply:
> We also need to update the tests for RSH8d to make sure that the updated value gets persisted.

(Same for [RSH8f](https://sdk.ably.com/builds/ably/specification/main/features/#RSH8f)).

Reply:
https://github.com/ably/ably-cocoa/pull/1847/commits/3098e3ae010a86aa37cf525af6b92f4a5b0ae079

Reply:
Thanks — leaving this comment unresolved so I remember to look at the test changes (will look at the tests once implementation comments addressed).